src/osd/PrimaryLogPG.cc

```diff
@@ -7899,7 +7899,8 @@ void PrimaryLogPG::write_update_size_and_usage(object_stat_sum_t& delta_stats, o
   } else if (length)
     ch.insert(offset, length);
   modified.union_of(ch);
-  if (write_full || offset + length > oi.size) {
+  if (write_full ||
```

Comment:
i don't think this issue exists in master. see https://github.com/ceph/ceph/commit/732a9504cbe050ce67bee28b7312fa5284b38084#diff-fb41013d27e932534adb50eb3de2aaa5R7790 

my guess is that the backend filestore did a short read, so the returned `result` was than `left`.  in the case of trimtrunc, this is expected. because the size of file representing the object is 0. so i think we should probably backport part of https://github.com/ceph/ceph/commit/732a9504cbe050ce67bee28b7312fa5284b38084#diff-fb41013d27e932534adb50eb3de2aaa5R7790 

@jdurgin what do you think?

Reply:
i exactly pinpoint the time sequence of the events that causing this bug , please follow my tracker link. 
In summary:
1. CEPHFS Client first issue a write op to the according object in the osd with truncate_seq 0,
the write was writeback cached in the Client side objectcacher for latter flush
2. latter on , client sent a truncate op to the mds, mds issued the truncate op with truncate_seq 2 to the same object.
3. on the osd side, it first receive the truncate op, set the oi.truncate_seq to 2, the truncate 
the according object to size 0.
4. the client side old write arrives (with truncate_seq 0), and its write offset start at 4063232,
with len for example 100. however, it found that the object has already got truncated to 0,
so shrink its write data buffer length to 0
5. finally, during write_update_size_and_usage, the code falsefully update the oi.size to the 
offset (4063232)

Reply:
the old previous write op, out of range of the truncated object, from the cephfs fuse client should not modify the oi.size. and persist the oi metadata onto the disk, since it caused mismatch among oi.size, oi.truncated_size and real object data size on the  backend. it is 
not relevant to whether the sparse async read or not. @jdurgin

Reply:
@tchaikov @hsepeng You're both right. The particular assert won't be triggered in luminous or later due to the changes to COPY_GET handling to update the cursor with the logical size, not the size read off disk. However, it doesn't make sense to change the logical object size when the write has been superseded by a trimtrunc.

Thus, I think this patch does make sense for master, even though it won't trigger the same assert in cache tiering. It fixes a flaw in the usage accounting for this trimtrunc + write race case. Also, it points out an area where we need more testing - trimtrunc handling.